oxide-auth-rocket/src/lib.rs

```diff
@@ -94,12 +96,13 @@ impl<'r> OAuthRequest<'r> {
     /// simplify the implementation of primitives and handlers, this type is the central request
     /// type for both these use cases. When you forget to provide the body to a request, the oauth
     /// system will return an error the moment the request is used.
-    pub fn add_body(&mut self, data: Data) {
+    pub async fn add_body(&mut self, data: Data<'_>) {
         // Nothing to do if we already have a body, or already generated an error. This includes
         // the case where the content type does not indicate a form, as the error is silent until a
         // body is explicitely requested.
         if let Ok(None) = self.body {
-            match serde_urlencoded::from_reader(data.open()) {
+            let data = data.open(2.mebibytes()).into_string().await;
```

Comment:
Arbitrary limit should be a constant somewhere, or, even better, be configurable on the object for the caller. Probably both where a default is set with a constant and it can be overwritten.